addons/link_tracker/tools/html_link_label_converter.py

```diff
@@ -0,0 +1,93 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from collections.abc import Callable
+
+import lxml
+from markupsafe import Markup
+
+MAX_LABEL_LENGTH = 40  # arbitrary
+
+
+class HTMLLinkLabelConverter:
```

Comment:
Have to say I am not a fan of this kind of implementation ... it makes the code hard to follow: you define functions (with Odoo code as there is a "create" a bit hidden in the 'map_func'), that are given to a converter that manipulates your data and modifies it in place ... Using a class is probably a matter of personal preferences (would have done it using tool functions as I don't think there is much added value using a complete class); but using callables hiding odoo calls and modifying the input is a bit meh to me. Especially it is used once in the code: general rule is to keep code simple and local, and extract it in a more generic way if used at least 3 times.

Don't have much proposals to do: I don't really like the whole approach of using a class. My opinion is that it tries to do smart code, while all we need is a small moulinette processing nodes on a small tool inside the link tracker module.

Reply:
Without going into detailed review (as the main point is which kind of implementation we are going to do), I would say that I would prefer something that 1/ finds the URLs (allowing to early find duplicates if any) 2/ creates link tacker in batch 3/ updates them. Current implementation is very "generic" (black box approach) while I expected something very "odoo-ish" (tailored for the need and to be improved when necessary), as anyway we use it once currently. Probably yet another big debate to have.

Reply:
I moved the code in the function body, and started implementing a required `search_or_create_mutli`.
Shall I continue with this version or are there structural problems here as well?